File: src/main/java/com/lessmarkup/interfaces/module/ModuleConfiguration.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.lessmarkup.interfaces.module;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.List;
import org.apache.commons.io.IOUtils;

public interface ModuleConfiguration {
    List<String> getElements();
    boolean isSystem();
    URL getUrl();
    String getModuleType();
    ClassLoader getClassLoader();
    ModuleInitializer getInitializer();
    InputStream getResourceAsStream(String path) throws IOException;
    default byte[] getResourceAsBytes(String path) throws IOException {
        try (InputStream stream = getResourceAsStream(path)) {
            return IOUtils.toByteArray(stream);
        }
    }
}

File: src/main/scala/com/lessmarkup/interfaces/module/ModuleConfiguration.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.interfaces.module

import java.io.{IOException, InputStream}
import java.net.URL

import org.apache.commons.io.IOUtils

trait ModuleConfiguration {
  def getElements: List[String]

  def isSystem: Boolean

  def getUrl: URL

  def getModuleType: String

  def getClassLoader: ClassLoader

  def getInitializer: ModuleInitializer

  @throws(classOf[IOException])
  def getResourceAsStream(path: String): InputStream

  @throws(classOf[IOException])
  def getResourceAsBytes(path: String): Array[Byte] = {
    val stream: InputStream = getResourceAsStream(path)
    try {
      IOUtils.toByteArray(stream)
    } finally {
      if (stream != null) stream.close()
    }
  }
}
